src/jit/morph.cpp

```diff
@@ -11876,12 +11876,31 @@ GenTreePtr Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* mac)
                 //
                 // a % b = a - (a / b) * b;
                 //
-                // NOTE: we should never need to perform this transformation when remorphing, since global morphing
-                //       should already have done so and we do not introduce new modulus nodes in later phases.
-                assert(!optValnumCSE_phase);
-                tree = fgMorphModToSubMulDiv(tree->AsOp());
-                op1  = tree->gtOp.gtOp1;
-                op2  = tree->gtOp.gtOp2;
+                // Use the suggested transform unless the special case transform works:
+                //
```

Comment:
Add this to the comment:
`// where b is a constant power of two.`

Reply:
Currently this allows for 1 and generates` (a & 0),` which probably should be be turned into just `a`

Reply:
It also seems like this optimization could  be used on other platforms besides Arm64

Reply:
@briansull 
> Currently this allows for 1 and generates(a & 0)

I would assume lowering/cse would be taking care of this elsewhere.  Does this belong in Morph?  It sounds rare.

> which probably should be be turned into just a

Maybe `0` but not `a`

> It also seems like this optimization could be used on other platforms besides Arm64

This is here to prevent disabling the optimization which is already in lowering.  So it is more of an undo a deoptimization, than an optimization per se.

The code would be slightly different for other platforms  
(Do not morph if `!op2->IsIntegralConst()`).

However, if I believe the comment in the `#else`, then there should be some slight CSE benefit for other platforms.

My guess is that it should replace the code in the `#else`, but I do not fully understand the `!isPow2(abs(divisor))` for GT_MOD below.

Options:
+ Leave it
+ Open an issue
+ Modify PR

@briansull Please advise.

Reply:
> However, if I believe the comment in the #else, then there should be some slight CSE benefit for other platforms.

Yep, if it weren't for CSE the best way to handle MOD would be to leave it as is until lowering and expand it there to SUB/DIV/MUL for ARM64. But it's not unusual to have code like:
```C#
int d = x / y;
int m = x % y;
```
Here it's preferable to expand MOD before CSE so that the resulting redundant DIV is eliminated.

It may be interesting to do this expansion for all platforms. However, CSE is not guaranteed and when it fails the SUB/DIV/MUL sequence is more expensive that MOD on platforms like x86/64 that do support MOD directly.

Reply:
>  but I do not fully understand the !isPow2(abs(divisor)) for GT_MOD below

That mirrors the condition from lowering. x/-2 is -(x/2) so it's still division by power of 2.

Reply:
@russellhadley Should we perform this optimization for all platforms?
when B is a constant power of two:

transform a%b to a&(b-1)

Reply:
Options:
@sdmaclea - You can open an issue for modifying this work so that all platform do this optimization.

Reply:
@briansull So the issue this was fixing was that since arm64 did not support GT_MOD/GT_UMOD, they were being universally eliminated in morph, this suppressed the a%b to a & (b-1) optimization which is already present for all platforms in `LowerUnsignedDivOrMod()`.  Therefore I think this is already done for all other platforms.

Reply:
OK, then all is fine.

Reply:
Given the amount of discussion here, and the somewhat confusing nature of "do this unless we believe that we can do better downstream", I would strongly urge more commenting here, especially with each assignment to `doMorph` below.